[DataPipe] Adding usage examples for IterDataPipes

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73033

Adding usage examples for IterDataPipes, with additional improvements for description of `groupby`, `IterDataPipe`, `MapDataPipe`.

Differential Revision: [D34313793](https://our.internmc.facebook.com/intern/diff/D34313793)

NivekT committed:
[DataPipe] Adding usage examples for IterDataPipes

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34313793).

ejguan commented on torch/utils/data/datapipes/iter/callable.py:
```diff
@@ -33,6 +33,18 @@ class MapperIterDataPipe(IterDataPipe[T_co]):
+        ...     return x + 1
+        >>> dp = IterableWrapper(range(10))
+        >>> map_dp_1 = dp.map(add_one)  # Invocation via functional form is preferred
+        >>> list(map_dp_1)  # [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```
nit: I think this is the actual output of this command
```suggestion
>>> list(map_dp_1)
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```

ejguan commented on torch/utils/data/datapipes/iter/combinatorics.py:
```diff
@@ -67,6 +67,14 @@ class ShufflerIterDataPipe(IterDataPipe[T_co]):
+        >>> import random
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> dp = IterableWrapper(range(10))
+        >>> random.seed(123)
```
Let's remove `seed` part for now. (I am not sure if we rely allow such seeding method in the future)

ejguan commented on torch/utils/data/datapipes/iter/callable.py:
```diff
@@ -33,6 +33,18 @@ class MapperIterDataPipe(IterDataPipe[T_co]):
+        >>> # We discourage the usage of `lambda` functions as they are not serializable with `pickle`
+        >>> # Use `functools.partial` or explicitly define the function instead
+        >>> map_dp_2 = Mapper(dp, lambda x: x + 1)
+        >>> list(map_dp_2)  # [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
```
nit: Ditto.

ejguan commented on torch/utils/data/datapipes/iter/combinatorics.py:
```diff
@@ -67,6 +67,14 @@ class ShufflerIterDataPipe(IterDataPipe[T_co]):
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> dp = IterableWrapper(range(10))
+        >>> random.seed(123)
+        >>> shuffle_dp = dp.shuffle()  # [0, 4, 1, 6, 3, 2, 9, 5, 7, 8]
```
Ditto

ejguan commented on torch/utils/data/datapipes/iter/combining.py:
```diff
@@ -363,6 +397,11 @@ class ZipperIterDataPipe(IterDataPipe[Tuple[T_co]]):
         *datapipes: Iterable DataPipes being aggregated
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> dp1, dp2, dp3 = IterableWrapper(range(5)), IterableWrapper(range(10, 15)), IterableWrapper(range(20, 25))
+        >>> list(dp1.zip(dp2, dp3))  # [(0, 10, 20), (1, 11, 21), (2, 12, 22), (3, 13, 23), (4, 14, 24)]
```
Not related to this diff. For syntax sugar, we may want to provide another functional api for such operation `torchdata.datapipes.iter.zip(dp1, dp2, dp3)`

ejguan commented on torch/utils/data/datapipes/iter/streamreader.py:
```diff
@@ -10,6 +10,12 @@ class StreamReaderIterDataPipe(IterDataPipe[Tuple[str, bytes]]):
         datapipe: Iterable DataPipe provides label/URL and byte stream
         chunk: Number of bytes to be read from stream per iteration.
             If ``None``, all bytes will be read util the EOF.
+    Example:
```
nit: Empty line above

ejguan reviewed (approved):
Thanks for adding it!

NivekT committed:
Update on "[DataPipe] Adding usage examples for IterDataPipes"


Adding usage examples for IterDataPipes, with additional improvements for description of `groupby`, `IterDataPipe`, `MapDataPipe`.

Differential Revision: [D34313793](https://our.internmc.facebook.com/intern/diff/D34313793)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34313793).

VitalyFedyunin commented on torch/utils/data/datapipes/iter/combinatorics.py:
```diff
@@ -67,6 +67,13 @@ class ShufflerIterDataPipe(IterDataPipe[T_co]):
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> dp = IterableWrapper(range(10))
+        >>> shuffle_dp = dp.shuffle()
+        [0, 4, 1, 6, 3, 2, 9, 5, 7, 8]
```
different order (of `list` and example output) in compare to the previous example
